### DISCUSSION TITLE
How to get current track position and duration?

### DISCUSSION THREAD

**Discussion by vitoyucepi** [savonet/liquidsoap#1822] [Q&A] [answered]
In official documentation there is `source.remaining()` function.

Is it possible to get track expected duration and current position(or progress)?
For example:
`source.duration(s)` -> `115.1`
`source.progress(s)` -> `12.29`
`source.remaining(s)` -> `102.81`

If there's no such options, then I can recreate this as feature request.

I've found this https://stackoverflow.com/questions/45672747/how-can-i-display-progress-and-track-length-in-the-video-stream-of-liquidsoap but it seems not very convenient.

**Comment by toots** [comment]
Historically, this is something that has been pretty terribly handled in liquidsoap for the following reasons:
* We first implemented file decoders as abstract callback-based decoders
* We did not have efficient ways of getting a file's `duration` from the decoders.

I just had a long look at the current implementation for decoders. It definitely needs to be cleaned. However, I am not sure if this will fit the current release. This is a pretty huge undertaking and we are at a point where we really want to stabilize the API to be able to reach something releasable.

For now, I would recommend using `request.duration` in conjunction with `source.remaining` to get the two other values. We will start working on a better implementation on our side as soon as `2.0.0` is out!

**Comment by toots** [comment, marked as answer]
Ok, I lied.. 😅 I found a simple way to add it: https://github.com/savonet/liquidsoap/commit/e9adaf11caf4e8c67a0f97a27d796e0789d92131.

Currently, `remaining` and `duration` are approximations, though. They may not be available at the very beginning of a track and will get more and more close to the actual value as we get close to the end of a track. `elapsed` should always be correct.
